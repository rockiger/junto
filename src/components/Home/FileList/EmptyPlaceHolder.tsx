import React from 'react'
import { Box } from '@material-ui/core'
import { styled } from '@material-ui/core/styles'
import { MdiReactIconComponentType } from 'mdi-react'

import { H2 } from 'components/gsuite-components/h2'

export { EmptyPlaceholder }

interface Props {
    icon?: MdiReactIconComponentType
    subline?: string
    title?: string
}

const EmptyPlaceholder = ({ icon, subline, title }: Props) => {
    const Icon = icon
    return (
        <Container>
            {icon && (
                //@ts-ignore
                <Icon
                    style={{
                        color: 'rgb(232, 240, 254)',
                        height: '10rem',
                        width: '10rem',
                    }}
                />
            )}
            <H2>{title}</H2>
            <div>{subline && subline}</div>
        </Container>
    )
}

const Container = styled(Box)({
    alignContent: 'center',
    alignItems: 'center',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    marginTop: 100,
})
