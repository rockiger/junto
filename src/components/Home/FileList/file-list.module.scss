@import '_base.scss';

.FileList_content {
    overflow-y: auto;
    height: calc((100vh - 65px) - 56px);
    @media (max-width: $md-max) {
        padding-left: 0.5rem;
    }
}

.FileList_tagline {
    margin-top: 1rem;
    font-weight: 500;
    font-size: 1rem;
}

.sortCriteria {
    @media (max-width: $md-max) {
        display: none;
    }
}
