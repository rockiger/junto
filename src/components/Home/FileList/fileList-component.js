// @ts-check

import React, { useGlobal } from 'reactn'
import { Link } from 'react-router-dom'
import { makeStyles } from '@material-ui/core/styles'
import { List, ListItem, ListItemIcon, ListItemText } from '@material-ui/core'
import FileDocumentIcon from 'mdi-react/FileDocumentIcon'
import SortAlphabeticalIcon from 'mdi-react/SortAlphabeticalIcon'

import Spinner from 'components/spinner'
import { EXT } from 'lib/constants'
import { getTitleFromFile, sortByDate } from 'lib/helper'
import { PageButtons } from 'components/pageButtons'
import { ButtonMenu } from 'components/ButtonMenu'

/** @typedef {{id: string, mimeType: string, name: string, modifiedByMeTime: string, trashed: boolean, viewedByMeTime: string}} File */
/** @typedef {'viewedByMeTime' | 'modifiedByMeTime' | 'sharedWithMeTime'} SortBy */

/**
 * @typedef {object} Props
 * @prop {File[]} files
 * @prop {SortBy} sortBy
 *
 */

/**
 *
 * @param {Props} props
 */
const FileListPartial = props => {
    const { files, sortBy } = props
    // @ts-ignore
    const [, setSearchTerm] = useGlobal('searchTerm')
    const classes = useStyles()
    return (
        <List className="filelist-list">
            {files
                .filter(file => {
                    return shouldFileDisplay(file)
                })
                .sort((file1, file2) => {
                    let result

                    if (sortBy === 'viewedByMeTime') {
                        result = sortByDate(
                            file1.viewedByMeTime,
                            file2.viewedByMeTime
                        )
                    } else {
                        result = sortByDate(
                            file1.modifiedByMeTime,
                            file2.modifiedByMeTime
                        )
                    }
                    return result
                })
                .map(file => {
                    const filename = getTitleFromFile(file)
                    return (
                        <ListItem className={classes.listitem} key={file.id}>
                            <Link
                                className={classes.link}
                                onClick={() => setSearchTerm('')}
                                to={`/page/${file.id}`}
                            >
                                <ListItemIcon className={classes.icon}>
                                    <FileDocumentIcon />
                                </ListItemIcon>
                                <ListItemText primary={filename} />
                            </Link>
                        </ListItem>
                    )
                })}
        </List>
    )
}

const PeriodList = ({ files, headline, sortBy }) => {
    if (files.length > 0) {
        return (
            <>
                <div className="filelist-tagline">{headline}</div>
                <FileListPartial files={files} sortBy={sortBy} />
            </>
        )
    } else {
        return null
    }
}

/**
 *
 * @param {{files: File[], sortBy: SortBy}} param0
 */

const Periods = ({ files, sortBy }) => {
    const createFilter = (older, younger = new Date()) => {
        return file => {
            // @ts-ignore
            const date = parseInt(Date.parse(file[sortBy]))

            return (
                date > parseInt(older.getTime()) &&
                // @ts-ignore
                date < parseInt(younger.getTime())
            )
        }
    }

    const newTimeBorder = (daysBack = 0) => {
        const today = new Date()
        today.setHours(0, 0, 0, 0)
        today.setDate(today.getDate() - daysBack)
        return new Date(today.getTime())
    }

    const today = newTimeBorder()
    const todayFilter = createFilter(today)
    const todayFiles = files.filter(todayFilter)

    const yesterday = newTimeBorder(1)
    const yesterdayFilter = createFilter(yesterday, today)
    const yesterdayFiles = files.filter(yesterdayFilter)

    const lastWeek = newTimeBorder(7)
    const lastWeekFilter = createFilter(lastWeek, yesterday)
    const lastWeekFiles = files.filter(lastWeekFilter)

    const lastMonth = newTimeBorder(30)
    const lastMonthFilter = createFilter(lastMonth, lastWeek)
    const lastMonthFiles = files.filter(lastMonthFilter)

    const earlier = new Date(0) // 1970
    const earlierFilter = createFilter(earlier, lastMonth)
    const earlierFiles = files.filter(earlierFilter)

    return (
        <>
            <PeriodList files={todayFiles} headline="Today" sortBy={sortBy} />
            <PeriodList
                files={yesterdayFiles}
                headline="Yesterday"
                sortBy={sortBy}
            />
            <PeriodList
                files={lastWeekFiles}
                headline="Previous 7 Days"
                sortBy={sortBy}
            />
            <PeriodList
                files={lastMonthFiles}
                headline="Previous 30 Days"
                sortBy={sortBy}
            />
            <PeriodList
                files={earlierFiles}
                headline="Earlier"
                sortBy={sortBy}
            />
        </>
    )
}

/**
 * @typedef {object} FileListComponentProps
 * @prop {string} [emptyMessage]
 * @prop {File[]} files
 * @prop {boolean} isLoading
 * @prop {SortBy} sortBy
 * @prop {string} searchTerm
 * @prop {function} setSortBy
 * @prop {string} title
 */

/**
 *
 * @param {FileListComponentProps} props
 */
const FileListComponent = props => {
    const { emptyMessage, files, searchTerm, setSortBy, sortBy, title } = props
    return (
        <div className="filelist">
            {setSortBy && (
                <PageButtons>
                    <strong style={{ fontWeight: 500, marginRight: '.5rem' }}>
                        {sortBy === 'viewedByMeTime'
                            ? 'Last opened by me'
                            : 'Last modified by me'}
                    </strong>
                    <ButtonMenu
                        items={[
                            {
                                key: 1,
                                name: 'Last modified by me',
                                handler: () => setSortBy('modifiedByMeTime'),
                                active: sortBy === 'modifiedByMeTime',
                            },
                            {
                                key: 2,
                                name: 'Last opened by me',
                                handler: () => setSortBy('viewedByMeTime'),
                                active: sortBy === 'viewedByMeTime',
                            },
                        ]}
                        selectable={true}
                    >
                        <SortAlphabeticalIcon />
                    </ButtonMenu>
                </PageButtons>
            )}
            <h1>{searchTerm ? 'Search Result' : title}</h1>
            <div className="filelist-content">
                {/* 
                // @ts-ignore */}
                {props.isLoading && <Spinner />}
                {!props.isLoading && <Periods files={files} sortBy={sortBy} />}
                {files.length === 0 && <h2>{emptyMessage}</h2>}
            </div>
            <style>{`
                    .filelist h1 {
                        border-bottom: 1px solid var(--border-color);
                        font-size: 1.5rem;
                        font-weight: 400;
                        margin: 0;
                        padding: .5rem;
                    }
                    .filelist-content {
                        overflow-y: auto;
                        height: calc((100vh - 65px) - 56px);
                    }
                    .filelist-tagline {
                        margin-top: 1rem;
                        font-weight: 500;
                        font-size: 1rem;
                    }
                    .filelist-list a {
                        border-radius: var(--border-radius);
                        color: var(--link-color);
                        font-size: 1rem;
                        padding: .5rem 1rem .5rem .75rem;
                    }
                    .filelist-list a:hover {
                        background-color: var(--hover-bg-color);
                        text-decoration: none;
                    }
                    .filelist-list a img, .filelist-list a span {
                        display: inline-block;
                        vertical-align: middle;
                    }
                `}</style>
        </div>
    )
}
export default FileListComponent

function useStyles() {
    const useStyles = makeStyles(theme => {
        return {
            icon: {
                color: theme.palette.primary.main,
                minWidth: theme.spacing(4),
            },
            link: {
                display: 'flex',
                flexGrow: 1,
                textDecoration: 'none',
                alignItems: 'center',
            },
            listitem: {
                padding: 0,
                paddingRight: theme.spacing(2),
            },
        }
    })
    return useStyles()
}

/**
 * @param {{mimeType: string, name: string, trashed: boolean}} file
 */
function shouldFileDisplay(file) {
    const { mimeType, name, trashed } = file
    return (
        mimeType === 'application/json' &&
        name.endsWith(EXT) &&
        trashed === false
    )
}
