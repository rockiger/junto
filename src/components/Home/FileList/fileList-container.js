// @ts-check
import React, { useEffect, useGlobal, useMemo } from 'reactn'
import { PageView } from 'components/Tracking'

import FileListComponent from './fileList-component'
import { sortByDate } from 'lib/helper'

export { FileList }
export default FileList

/** @typedef {'viewedByMeTime' | 'modifiedByMeTime' | 'sharedWithMeTime'} SortBy */

/**
 * @typedef FileListProps
 * @property {any} [emptyIcon]
 * @property {string} [emptyMessage]
 * @property {string} [emptySubline]
 * @property {'h1'|'h2'|'h3'|'h4'|'h5'|'h6'} [header]
 * @property {boolean} [isLoading]
 * @property {boolean} [isScrollable]
 * @property {import('reactn/default').IFile[]} files
 * @property {SortBy} [sortBy]
 * @property {()=>{}} [setSortBy]
 * @property {string} [title]
 */

/**
 * @param {FileListProps} props
 */
function FileList({
    emptyIcon,
    emptyMessage,
    emptySubline,
    files,
    header,
    sortBy,
    setSortBy,
    title,
}) {
    const [isFileListLoading] = useGlobal('isFileListLoading')
    const [searchTerm] = useGlobal('searchTerm')

    useEffect(() => PageView(), [])

    const filePeriods = useMemo(() => {
        console.log('filePeriods', files, sortBy)
        if (_.isEmpty(files)) {
            return {
                todayFiles: [],
                yesterdayFiles: [],
                lastWeekFiles: [],
                lastMonthFiles: [],
                earlierFiles: [],
            }
        }
        const sorter = createSorter(sortBy)
        const sortedFiles = files.sort(sorter)
        const today = newTimeBorder()
        const todayFilter = createFilter(today)
        const todayFiles = sortedFiles.filter(todayFilter)

        const yesterday = newTimeBorder(1)
        const yesterdayFilter = createFilter(yesterday, today)
        const yesterdayFiles = sortedFiles.filter(yesterdayFilter)

        const lastWeek = newTimeBorder(7)
        const lastWeekFilter = createFilter(lastWeek, yesterday)
        const lastWeekFiles = sortedFiles.filter(lastWeekFilter)

        const lastMonth = newTimeBorder(30)
        const lastMonthFilter = createFilter(lastMonth, lastWeek)
        const lastMonthFiles = sortedFiles.filter(lastMonthFilter)

        const earlier = new Date(0) // 1970
        const earlierFilter = createFilter(earlier, lastMonth)
        const earlierFiles = sortedFiles.filter(earlierFilter)
        return {
            todayFiles,
            yesterdayFiles,
            lastWeekFiles,
            lastMonthFiles,
            earlierFiles,
        }
    }, [files, sortBy])

    return (
        <FileListComponent
            emptyIcon={emptyIcon}
            emptyMessage={
                searchTerm
                    ? 'None of your files matched this search.'
                    : emptyMessage
            }
            emptySubline={emptySubline}
            filePeriods={filePeriods}
            header={header}
            isLoading={_.isEmpty(files) && isFileListLoading}
            searchTerm={searchTerm}
            setSortBy={setSortBy}
            //@ts-ignore
            sortBy={sortBy}
            title={title}
        />
    )
}

function createFilter(older, younger = new Date(), sortBy = 'modified') {
    return file => {
        // @ts-ignore
        const date = parseInt(Date.parse(file[sortBy]))

        return (
            date > parseInt(older.getTime()) &&
            // @ts-ignore
            date < parseInt(younger.getTime())
        )
    }
}

function newTimeBorder(daysBack = 0) {
    const today = new Date()
    today.setHours(0, 0, 0, 0)
    today.setDate(today.getDate() - daysBack)
    return new Date(today.getTime())
}

function createSorter(sortBy = 'modified') {
    return (file1, file2) => {
        return sortByDate(file1[sortBy], file2[sortBy])
    }
}
