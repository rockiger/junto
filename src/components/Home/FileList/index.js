import FileList from './fileList-container'

export default FileList
export { FileList }
