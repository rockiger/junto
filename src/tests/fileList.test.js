import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import FileList from '../components/fileList';

describe('FileList', () => {
    const props = {
        setIsSigningIn: (isSignedIn) => {},
    };

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(
            <FileList />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    test('has a valid snapshot', () => {
        const component = renderer.create(
            <FileList />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

});