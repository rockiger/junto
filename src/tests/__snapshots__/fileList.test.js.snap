// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileList has a valid snapshot 1`] = `
<div
  className="filelist"
>
  <h1>
    Your work
  </h1>
  <div
    className="filelist-tagline"
  >
    Last edited
  </div>
  <div
    className="Spinner"
  >
    <svg
      className="Spinner-spinner"
      height="32px"
      viewBox="0 0 66 66"
      width="32px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="Spinner-path"
        cx="33"
        cy="33"
        fill="none"
        r="30"
        strokeLinecap="round"
        strokeWidth="6"
      />
    </svg>
    <style>
      
            .Spinner {
                width: 100%;
                height: 100%;
                display: flex;
                justify-content: center;
                padding-top: 2rem;
            }
            .Spinner-spinner {
            animation: rotator 1.4s linear infinite;
            }
            @keyframes rotator {
            0% {
                transform: rotate(0deg);
            }
            100% {
                transform: rotate(270deg);
            }
            }
            .Spinner-path {
            stroke-dasharray: 187;
            stroke-dashoffset: 0;
            transform-origin: center;
            animation: dash 1.4s ease-in-out infinite, colors 5.6s ease-in-out infinite;
            }
            @keyframes colors {
            0% {
                stroke: #4285f4;
            }
            25% {
                stroke: #de3e35;
            }
            50% {
                stroke: #f7c223;
            }
            75% {
                stroke: #1b9a59;
            }
            100% {
                stroke: #4285f4;
            }
            }
            @keyframes dash {
            0% {
                stroke-dashoffset: 187;
            }
            50% {
                stroke-dashoffset: 46.75;
                transform: rotate(135deg);
            }
            100% {
                stroke-dashoffset: 187;
                transform: rotate(450deg);
            }
            }                
        
    </style>
  </div>
  <style>
    
                        .filelist h1 {
                            border-bottom: 1px solid var(--border-color);
                            font-size: 1.5rem;
                            font-weight: 400;
                            margin: 0;
                            padding: 1rem .5rem;
                        }
                        .filelist-list {
                            list-style-type: none;
                            padding-inline-start: unset;
                        }
                        .filelist-tagline {
                            margin-top: 1rem;
                            font-weight: 600;
                        }
                        .filelist-list a {
                            border-radius: var(--border-radius);
                            color: var(--link-color);
                            display: block;
                            font-size: 1rem;
                            padding: .5rem 1rem;
                        }
                        .filelist-list a:hover {
                            background-color: var(--hover-bg-color);
                            text-decoration: none;
                        }
                        .filelist-list a img {
                            height: 1rem;
                            width: 1rem;
                            padding-right: 1rem;
                        }
                        .filelist-list a img, .filelist-list a span {
                            display: inline-block;
                            vertical-align: middle;
                        }
                    
  </style>
</div>
`;
